### PR TITLE
Enable Brotli decompression

### LINKED ISSUE #43392: Quarkus does not decompress Brotli requests in native mode
### Describe the bug

I have an application which uses brotli for compression of HTTP requests and responses. It works as expected in JVM mode, but in native it fails to decrypt incoming requests.

The method in question looks like this:

```
    @POST
    @Path("/decompression")
    @Produces(MediaType.TEXT_PLAIN)
    public String decompressionHttpResponse(byte[] compressedData) {
        String decompressedData = new String(compressedData);
        System.out.println(decompressedData);
        return decompressedData;
    }
```

### Expected behavior

Brotli should decompress the input and return the decompressed data (like it does in JVM)

### Actual behavior

Brotli returns compressed data

### How to Reproduce?

1. `git clone -b brotli-native https://github.com/fedinskiy/reproducer`
2. `cd reproducer`
3. `mvn clean verify -Dnative` # GreetingResourceTest, which is run in JVM succeeds, GreetingResourceIT, which is run the same checks in native fails

### Output of `uname -a` or `ver`

6.10.4-200.fc40.x86_64

### Output of `java -version`

Java version: 21.0.1, vendor: Eclipse Adoptium

### Quarkus version or git rev

3.14.3

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.6 (bc0240f3c744dd6b6ec2920b3cd08dcc295161ae)

### Additional information

This issue is a follow up to https://github.com/quarkusio/quarkus/issues/40533, which used the same cases with slightly different properties

### PR DESCRIPTION
**Pull request by zakkak**
I tried adding a test that gets an encrypted text and sends it encrypted to the server for decryption as in https://github.com/quarkusio/quarkus/issues/43392 with no success (I couldn't get the test properly set up, not that it decompression fails). Manually running the reproducer from https://github.com/quarkusio/quarkus/issues/43392 confirms the fix, but we should have a test in Quarkus IT. @Karm do you think you can help with this given your prior experience?

Note/Thoughts: Zstd and Snappy (cc @gastaldi) remain unsupported. Since we already support snappy in Kafka, and there is quarkus-snappy we should probably support it in vertx as well.

Fixes https://github.com/quarkusio/quarkus/issues/43392

### CODE REVIEW

**Review by cescoffier — APPROVED** [review]

**Comment by geoand** [comment]
Are we waiting for something for this one or should it go in?

**Comment by Karm** [comment]
@geoand I'll add a test case, hopefully today. Could be a separate PR, could be this one...

**Comment by geoand** [comment]
Thanks for the update. I'll ket @zakkak comment on his preference

**Comment by zakkak** [comment]
Given that adding a test is in @Karm's immediate plans I prefer waiting for it so that we can backport the PR as a whole (including tests) in one go.

**Comment by Karm** [comment]
@zakkak @cescoffier I added a Decompressor tests in commit [1aa2223](https://github.com/quarkusio/quarkus/pull/44348/commits/1aa22234b20c1cd27d624f59ddd823f4839ab430)

It works fine with HotSpot, client sends compressed PUT and POST to various endpoints and Server decompresses and either sends back or compresses again and sends back, client decompresses and checks. All fine.

## Native testing fails

```
 ./mvnw clean verify -f integration-tests -pl  vertx-http-compressors/app,vertx-http-compressors/all,vertx-http-compressors/some -Dnative
```

Native test flow fails for a peculiar reason: 

```
java.lang.UnsatisfiedLinkError: 'java.nio.ByteBuffer com.aayushatharva.brotli4j.encoder.EncoderJNI.nativeCreate(long[])'
        at com.aayushatharva.brotli4j.encoder.EncoderJNI.nativeCreate(Native Method)
        at com.aayushatharva.brotli4j.encoder.EncoderJNI.access$200(EncoderJNI.java:17)
        at com.aayushatharva.brotli4j.encoder.EncoderJNI$Wrapper.<init>(EncoderJNI.java:90)
        at com.aayushatharva.brotli4j.encoder.Encoder.<init>(Encoder.java:48)
        at com.aayushatharva.brotli4j.encoder.BrotliOutputStream.<init>(BrotliOutputStream.java:36)
        at com.aayushatharva.brotli4j.encoder.BrotliOutputStream.<init>(BrotliOutputStream.java:48)
        at com.aayushatharva.brotli4j.encoder.BrotliOutputStream.<init>(BrotliOutputStream.java:58)
        at io.quarkus.compressors.it.Testflow.compress(Testflow.java:153)
        at io.quarkus.compressors.it.Testflow.runDecompressorsTest(Testflow.java:125)
        at io.quarkus.compressors.it.RESTEndpointsTest.testDecompressors(RESTEndpointsTest.java:120)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at io.quarkus.test.junit.QuarkusTestExtension.interceptTestTemplateMethod(QuarkusTestExtension.java:856)
```

I originally thought that for some reason the test code itself is baked in the executable, but it is not the case. 

So far it escapes me what happens that when running the integration test native, the *test code*, i.e. what is typed in the Junit, cannot find the Brotli4j .so :-(

**Comment by Karm** [comment]
For the record, doing this doesn't help the native integration test:
```
--- a/integration-tests/vertx-http-compressors/app/pom.xml
+++ b/integration-tests/vertx-http-compressors/app/pom.xml
@@ -21,6 +21,11 @@
             <artifactId>quarkus-junit5</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>com.aayushatharva.brotli4j</groupId>
+            <artifactId>brotli4j</artifactId>
+            <scope>test</scope>
+        </dependency>
         <dependency>
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-rest</artifactId>

```
and neither does this:
```
--- a/integration-tests/vertx-http-compressors/all/pom.xml
+++ b/integration-tests/vertx-http-compressors/all/pom.xml
@@ -29,6 +29,11 @@
             <type>test-jar</type>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>com.aayushatharva.brotli4j</groupId>
+            <artifactId>brotli4j</artifactId>
+            <scope>test</scope>
+        </dependency>
         <dependency>
             <groupId>io.vertx</groupId>
             <artifactId>vertx-web-client</artifactId>
```

**Comment by Karm** [comment]
I'll revisit it later...

**Comment by Karm** [comment]
The CI test result is apparently misleading again 😔 I haven't fixed the Native test yet. It passed for me locally only with HotSpot.

**Comment by Karm** [comment]
@zakkak @cescoffier 
This "fixes" the native test:

```diff
--- a/integration-tests/vertx-http-compressors/app/src/test/java/io/quarkus/compressors/it/Testflow.java
+++ b/integration-tests/vertx-http-compressors/app/src/test/java/io/quarkus/compressors/it/Testflow.java
@@ -149,6 +149,7 @@ public static Buffer compress(String algorithm, String payload) {
             }
             return Buffer.buffer(byteStream.toByteArray());
         } else if ("br".equalsIgnoreCase(algorithm)) {
+            System.load("/tmp/brotli/lib/linux-x86_64/libbrotli.so");
             try (BrotliOutputStream brotliStream = new BrotliOutputStream(byteStream)) {
                 brotliStream.write(payload.getBytes(StandardCharsets.UTF_8));
             } catch (IOException e) {
```

Do you think it's some kind of classloading issue that manifests only when running with `-Dnative`, the way the `all` and `some` apps depend on the `app` module?

**Comment by Karm** [comment]
I have a fix, gonna push later.
One needs to call `Brotli4jLoader.ensureAvailability();` to load the .so. Obviously...   The fact that you don't have to do it in HotSpot testing as QuarkusTest is that you run it all in the same instance and the brotli is laoded already by Q.

**Review by Karm — APPROVED** [review]
I added tests and it's all good now. Thanks.

**Comment by Karm** [comment]
https://github.com/quarkusio/quarkus/issues/44422

**Review by cescoffier — APPROVED** [review]

File: extensions/netty/runtime/src/main/java/io/quarkus/netty/runtime/graal/NettySubstitutions.java [hunk 1]
```diff
@@ -533,21 +538,23 @@ final class Target_io_netty_handler_codec_http2_DelegatingDecompressorFrameListe
...
-        } else {
-            return new EmbeddedChannel(ctx.channel().id(), ctx.channel().metadata().hasDisconnect(), ctx.channel().config(),
-                    new ChannelHandler[] { ZlibCodecFactory.newZlibDecoder(ZlibWrapper.GZIP) });
+        if (GZIP.contentEqualsIgnoreCase(contentEncoding) || X_GZIP.contentEqualsIgnoreCase(contentEncoding)) {
+            return new EmbeddedChannel(ctx.channel().id(), ctx.channel().metadata().hasDisconnect(),
+                    ctx.channel().config(), ZlibCodecFactory.newZlibDecoder(ZlibWrapper.GZIP));
         }
+        if (DEFLATE.contentEqualsIgnoreCase(contentEncoding) || X_DEFLATE.contentEqualsIgnoreCase(contentEncoding)) {
+            final ZlibWrapper wrapper = strict ? ZlibWrapper.ZLIB : ZlibWrapper.ZLIB_OR_NONE;
+            // To be strict, 'deflate' means ZLIB, but some servers were not implemented correctly.
```
[the remark this answers is not in the document]
Comment (zakkak) [reply extensions/netty/runtime/src/main/java/io/quarkus/netty/runtime/graal/NettySubstitutions.java:547]: @Karm that was copy-pasted from upstream, see https://github.com/netty/netty/blob/7b4fe3d25c01576636985320ebf16393cd21c13c/codec-http2/src/main/java/io/netty/handler/codec/http2/DelegatingDecompressorFrameListener.java#L180 (and it looks like it's 10 years old)

We could have skipped it here, but I just went with copy-paste and removing only what we don't need.

**Comment by zakkak** [comment]
Thank you @Karm !!!

**Pull request merged**